### DISCUSSION TITLE
Training Yolo on a custom 3 class dataset

### DISCUSSION THREAD

**Discussion by TeddyxDaou** [ultralytics/ultralytics#25669] [Q&A] [answered]
Training YOLOv9-C on a custom 3-class dataset (batch 16, imgsz 640, AdamW), mAP50 plateaus at 0.71 after epoch 80 despite loss still decreasing on both box and cls terms. Is this a sign of label noise, or is box loss just converging slower than cls at this stage?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @TeddyxDaou, thank you for your interest in Ultralytics 🚀! This is an automated response, and an Ultralytics engineer will also assist you soon.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for new users, where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many common questions may already be answered.

For this custom training ❓ question, please provide as much information as possible, including:
- Dataset image examples and representative labels
- Your dataset YAML configuration
- Complete training command or Python code
- Training and validation logs, including metric plots
- Ultralytics and PyTorch versions
- Hardware and environment details
- Any validation or prediction examples showing incorrect detections

Please also verify that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

If this is a 🐛 bug report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package, including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments, with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by fadyabirached** [comment, marked as answer]
These aren't mutually exclusive, and I'd check per-class AP first before deciding between them; if one class has much lower AP than the others, that points more toward label noise/class imbalance rather than pure convergence speed. If box loss is still dropping meaningfully while cls loss has basically flattened, that's a pretty normal decoupling since box regression usually takes longer to converge than classification, so I wouldn't read that alone as a problem. Also worth checking if val mAP and train mAP are both plateauing together or diverging, since a growing gap there would point more toward overfitting/noise than pure slow convergence.

**Reply by TeddyxDaou** [reply]
Really appreciate this, thanks! I was overthinking the loss curves, and your explanation helped me understand what I should actually be looking at. This cleared things up a lot.
